### ISSUE TITLE
Use block-level elements consistently

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#602]
I would like to propose the implementation and inclusion of a new StyleCop diagnostic.

**Category:** Documentation
**Rule:** Use block-level elements consistently
**Description:** If any element of the XML documentation contains a block-level element as a child, then all immediate children of the element should be block-level elements. This generally applies to elements containing multiple paragraphs (the `<para>` block-level element), but can also be affected by elements like `<list>` and `<note>`.
**Note:** This is an expansion of #601.

**Rationale:** This rule ensures these elements appear consistently when rendered to a variety of output media, including HTML.

**Example:** The most common violation of this rule in my own documentation of the past is the following:

```csharp
/// <exception cref="ArgumentNullException">
/// If <paramref name="x"/> is <see langword="null"/>.
/// <para>-or-</para>
/// <para>If <paramref name="y"/> is <see langword="null"/>.</para>
/// </exception>
```

Since the `<exception>` element contained at least one `<para>` element as a child, the first sentence should also have been written within a `<para>` element:

```csharp
/// <exception cref="ArgumentNullException">
/// <para>If <paramref name="x"/> is <see langword="null"/>.</para>
/// <para>-or-</para>
/// <para>If <paramref name="y"/> is <see langword="null"/>.</para>
/// </exception>
```

**Comment by AArnott** [comment]
Does this mean that if one `param` has a `para` tag in it, that all `param` tags need one? I hope not.

**Comment by sharwell** [comment]
It does. Again, a code fix is already implemented that provides this consistency without making the developer do a bunch of work.

**Comment by AArnott** [comment]
It's not so much the amount of work I'm worried about. It's readability of the doc comments in their original xml form. They were obviously designed to be minimal. Forcing `<para>` tags to be present when they don't add any value just adds extraneous xml making it harder to read. 
My expectation of StyleCop is that it *improves* readability. How does requiring extra xml tags improve readability?

**Comment by sharwell** [comment]
I just realized that we're talking in the context of the wrong rule. Issue #603 covers the specific case which you asked about.

All three rules for `<para>` elements (and their code fixes) were originally created to simplify the task of ensuring the final rendered documentation is as high-quality as possible without requiring more work than necessary from the developer. The added value comes from enabling the CSS files associated with HTML output to consistently style content for parameters. When some content is placed in `<para>` elements (which become `<p>` elements in HTML), and other content is not, it tends to produce noticeable inconsistencies in the final pages.

In my experience, the rule proposed in #603 did not commonly occur for `<param>` or `<item>` elements, but occurred several times for `<exception>` elements.

**Comment by sharwell** [comment]
@DotNetAnalyzers/members This rule really helps ensure consistent formatting when using tools like Sandcastle Help File Builder to publish documentation, and manually checking for these cases is tiresome. Since the rule includes a code fix that can correct existing violations, I believe it would be a great help to developers if we could include it.

**Comment by Noryoko** [comment]
:+1:

**Comment by otac0n** [comment]
:+1: I feel this one fits better than #601

**Comment by sharwell** [comment]
I'm going to move this to a new project.
